File: PriceCalculatorService.ts
import { Cake } from "./Cake";
import { Dough, Fillings } from "./types";

export class PriceCalculatorService {
  calculateTotalPrice(cake: Cake): number {
    const basePrice = cake.getBasePrice();
    const modifiers = cake.getPriceModifiers();

    let total = basePrice;

    // Dough modifier
    if (cake.dough === Dough.RED_VELVET) {
      total += modifiers.redVelvet;
    }

    // Pistachio modifier
    if (cake.fillings.some(filling => filling === Fillings.PISTACHIO_CREAM)) {
      total += modifiers.pistachio;
    }

    // Toppings modifiers (up to 2, handled in the ui)
    total += cake.toppings.length * modifiers.topping;

    return total;
  }
}

File: PoC/src/PriceCalculatorService.java
package src;

public class PriceCalculatorService {
  public double calculateTotalPrice(Cake cake) {
    double basePrice = cake.getBasePrice();
    PriceModifiers modifiers = cake.getPriceModifiers();

    double total = basePrice;

    if (cake.dough == Dough.RED_VELVET) {
      total += modifiers.getRedVelvet();
    }

    if (cake.fillings.contains(Fillings.PISTACHIO_CREAM)) {
      total += modifiers.getPistachio();
    }

    total += cake.toppings.size() * modifiers.getTopping();

    return total;
  }
}
